### ISSUE TITLE
Prototyping `dallinger develop` command with PsyNet: 5pts

### ISSUE THREAD

**Issue by pmcharrison** [Dallinger/Dallinger#3337]
The `dallinger develop` command is very promising and looks like it will save us a lot of time in the future.

When applying it to our own experiments we do have some problems though and they are quite subtle for us to debug. We were hoping that @jessesnyder could help here by trying to run the code himself.

As a starting point, it would be great to trial one of the PsyNet demos, for example `static`. The PsyNet repository can be found [here](https://gitlab.com/computational-audition-lab/psynet). 

Ordinarily I would run this demo by navigating to the demo directory (`demos/static`) and run `dallinger debug`. Here I instead follow the Dallinger develop instructions as specified [here](https://github.com/Dallinger/Dallinger/pull/3161).

The first issue I get is an import error, that doesn't occur with `dallinger debug`.

```
(psynet) peter@dhcp-10-248-39-251 develop % dallinger develop launch

❯❯ Error accessing http://127.0.0.1:5000/launch (500):
{"status": "error", "message": "Failed to load experiment in /launch: No module named 'psynet.experiment'"}
```

This error comes from the top of `experiment.py`:

```
# pylint: disable=unused-import,abstract-method

##########################################################################################
# Imports
##########################################################################################

import logging

from flask import Markup

import psynet.experiment
```

It turns out one can sidestep this issue by copying the `static` directory outside the PsyNet source code tree and rerunning the developer commands from there. So, this issue is not so crucial, but it would be nice to solve.

When I do this, I can get a few pages through, until 'Do you want to enable custom stimulus and stimulus version filters?'. I then run into a more serious error that I have been unable to debug yet.

```
INFO:werkzeug:127.0.0.1 - - [17/Sep/2021 15:00:14] "POST /response HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/peter/git/psynet/psynet/experiment.py", line 946, in route_response
    res = exp.process_response(
  File "/Users/peter/git/psynet/psynet/experiment.py", line 630, in process_response
    self.timeline.advance_page(self, participant)
  File "/Users/peter/git/psynet/psynet/timeline.py", line 1329, in advance_page
    new_elt.consume(experiment, participant)
  File "/Users/peter/git/psynet/psynet/timeline.py", line 230, in consume
    target_elt = self.get_target(experiment, participant)
  File "/Users/peter/git/psynet/psynet/timeline.py", line 264, in get_target
    val = call_function(
  File "/Users/peter/git/psynet/psynet/utils.py", line 107, in call_function
    return function(*arg_values)
  File "/Users/peter/git/psynet/psynet/trial/main.py", line 1533, in <lambda>
    lambda experiment, participant: self._get_current_trial(participant)
  File "/Users/peter/git/psynet/psynet/trial/main.py", line 1489, in _get_current_trial
    return self.trial_class.query.get(trial_id)
  File "<string>", line 2, in get
    
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 390, in warned
    return fn(*args, **kwargs)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 944, in get
    return self._get_impl(ident, loading.load_on_pk_identity)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 948, in _get_impl
    return self.session._get_impl(
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2839, in _get_impl
    return db_load_fn(
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 568, in load_on_pk_identity
    return result.one()
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1374, in one
    return self._only_one_row(
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 558, in _only_one_row
    row = onerow(hard_close=True)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1238, in _fetchone_impl
    return self._real_result._fetchone_impl(hard_close=hard_close)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 1626, in _fetchone_impl
    row = next(self.iterator, _NO_ROW)
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 150, in chunks
    rows = [proc(row) for row in fetch]
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 150, in <listcomp>
    rows = [proc(row) for row in fetch]
  File "/Users/peter/opt/miniconda3/envs/psynet/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 1300, in polymorphic_instance
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Row with identity key (<class 'melody.psychophysics.RatingPitchIntervalTrial'>, (2,), None) can't be loaded into an object; the polymorphic discriminator column 'info.type' refers to mapped class RatingPitchIntervalPractiseTrial->info, which is not a sub-mapper of the requested mapped class RatingPitchIntervalPractiseTrial->info
```

The responsible line of code is this:

``` py
return self.trial_class.query.get(trial_id)
```

The `trial_class` is defined in `experiment.py`. I can successfully retrieve this `Trial` object if I instead use a class defined in the `psynet` package: 

``` py
return Trial.query.get(trial_id)
```

Looking at the classes of these objects:

``` py
Trial.query.get(trial_id).__class__
# <class 'dallinger_experiment.experiment.AnimalTrial'>

self.trial_class
# <class 'develop.experiment.AnimalTrial'>
```

Somehow `AnimalTrial` is getting registered in SQLAlchemy under `dallinger_experiment.experiment.AnimalTrial` instead of under `develop.experiment.AnimalTrial`, but I'm not sure how/why this is happening. I hope maybe @jessesnyder can figure it out!

PS I tried to reproduce the equivalent problem with the Dallinger MCMCP demo but had no luck.

**Comment by jessesnyder** [comment]
@pmcharrison I think I have the first problem ("No module named 'psynet.experiment'") figured out, with a workaround, rather than a real resolution. I had the same error, which I traced to a typically complicated chain of events during import.

## Diagnostics

1. `flask run` imports /static/develop/app.py, and this adds a bunch of keys to `sys.modules`:
```
import sys; print("\n".join(sorted([mod for mod in sys.modules.keys() if mod.startswith("psynet")])))
psynet
psynet.demos
psynet.demos.static
psynet.demos.static.develop
psynet.demos.static.develop.app
```
For me, these all begin with "psynet" because that's the name of the folder I created when cloning the repository from gitlab, and for some reason I haven't figured out, the _parent_ of that folder is in `sys.path` ('/Users/jesses/projects/py_dev/dlgr_demos').

At this point, `psynet` is in `sys.modules`, but subpackages (like `psynet.experiment`) are **not**.

2. When `import psynet.experiment` runs in experiment.py, I believe the import machinery thinks "we've already loaded a path for `psynet`", so it doesn't go to `sys.path` and the filesystem to look for a module called `psynet.experiment`. Instead, it just looks in the existing mapping under the `psynet` key, and there's no attribute called `experiment` there, so it bombs.

## Workaround
I renamed my git clone folder `psynet_stuff`, and then it worked, because `flask run` no longer puts `psynet` into `sys.modules`; it puts `psynet_stuff` there instead:
```
import sys; print("\n".join(sorted([mod for mod in sys.modules.keys() if mod.startswith("psynet")])))
psynet_stuff
psynet_stuff.demos
psynet_stuff.demos.static
psynet_stuff.demos.static.develop
psynet_stuff.demos.static.develop.app
```
This means that when `import psynet.experiment` executes, it finds no existing keys in `sys.modules`, and so consults `sys.path`, and finds the correct file to load as a module.

**Comment by jessesnyder** [comment]
Of possible relevance, from @MatthewWilkes's book:

![image](https://user-images.githubusercontent.com/63560/134746973-64979fa8-5ebe-4e4e-8ca1-d518d4ee5db0.jpeg)

**Comment by pmcharrison** [comment]
Thanks, @jessesnyder -- looks like you made useful progress here!

**Comment by pmcharrison** [comment]
Also the book looks cool. Hope to check it out sometime!

**Comment by jessesnyder** [comment]
> Also the book looks cool. Hope to check it out sometime!

It deserved heavier paper stock. ;-)

**Comment by MatthewWilkes** [comment]
@jessesnyder Try the Russian translation, it's a hardback and nicer paper ;)

**Comment by alecpm** [comment]
@jessesnyder Is this something that would be (partly) solved by a `pip install -e .` in psynet?

**Comment by jessesnyder** [comment]
> @jessesnyder Is this something that would be (partly) solved by a `pip install -e .` in psynet?

@alecpm Did you mean installing psynet that way, or installing dallinger that way from psynet?

**Comment by alecpm** [comment]
I meant installing psynet that way, to help ensure `import psynet.experiment` works even when not in the psynet parent dir.

**Comment by jessesnyder** [comment]
@alecpm Interestingly, this seems to make no difference. The presence of a directory named `psynet` under the current working directory path winds up adding the `psynet` name as a key to `sys.modules`, and once that happens, import never consults `sys.path` again to find it. The problem seems to stem from the fact that for some reason running `flask run` adds the folder containing the `psynet` directory to `sys.path`. I have no idea why; it's not the directory I'm calling `flask run` from.

From `flask run`:
```
(Pdb) print("\n".join(sys.path))
/Users/jesses/projects/py_dev/dlgr_demos  **<-- why, oh why, are you here?**
/Users/jesses/projects/py_dev/dlgr_demos/psynet/demos/static/.venv/bin
/Users/jesses/.pyenv/versions/3.9.5/lib/python39.zip
/Users/jesses/.pyenv/versions/3.9.5/lib/python3.9
/Users/jesses/.pyenv/versions/3.9.5/lib/python3.9/lib-dynload
/Users/jesses/projects/py_dev/dlgr_demos/psynet/demos/static/.venv/lib/python3.9/site-packages
/Users/jesses/projects/py_dev/Dallinger3
/Users/jesses/projects/py_dev/dlgr_demos/psynet
```
Invoking python directly:
```
>>> print("\n".join(sys.path))

/Users/jesses/.pyenv/versions/3.9.5/lib/python39.zip
/Users/jesses/.pyenv/versions/3.9.5/lib/python3.9
/Users/jesses/.pyenv/versions/3.9.5/lib/python3.9/lib-dynload
/Users/jesses/projects/py_dev/dlgr_demos/psynet/demos/static/.venv/lib/python3.9/site-packages
/Users/jesses/projects/py_dev/Dallinger3
/Users/jesses/projects/py_dev/dlgr_demos/psynet
```

**Comment by jessesnyder** [comment]
I got to the cause of the problem, which is in `flask.cli.prepare_import`; specifically:
```python
    # move up until outside package structure (no __init__.py)
    while True:
        path, name = os.path.split(path)
        module_name.append(name)
```
In my case, this adds the directory _containing_ the psynet root folder to `sys.path`, precipitating the subsequent problems documented above.

I looked into an [alternate method](https://flask.palletsprojects.com/en/2.0.x/cli/?#custom-scripts) of starting Flask that would avoid the find-the-app-module-and-import-it process, using a custom click command which specifies a `create_app` app factory, but I don't like the warning about breaking the reloader:
> When using a custom script, if you introduce an error in your module-level code, the reloader will fail  because it can no longer load the entry point.
 The flask command, being separate from your code, does not have this issue and is recommended in most cases.

**Comment by pmcharrison** [comment]
Well debugged, that sounds like a serious investigation!

I wonder if we should just make a note of this and move on? It seems like there's not much we can do here, right? Shall we just rename the Git cloned folder? Would making it `PsyNet` instead of `psynet` be sufficient?

I think it would be good to try and get onto that second issue, because that was where I got really stuck.

**Comment by jessesnyder** [comment]
I think that renaming would be sufficient, since `sys.modules` is a `dict`, and `dict` keys are case sensitive.

I'm less concerned about the specific error, and more concerned about the whole category of errors that are likely to recur in subtly different guises due to the manipulation of `sys.path` from `flask run`. I don't see a way around this problem, so for now I agree we just need to treat it as a known limitation.

**Comment by pmcharrison** [comment]
I see. Ok then!

**Comment by jessesnyder** [comment]
@pmcharrison From the Ambiguous Cause/Effect Department... I started Flask without starting a worker or clock process, and the SQLAlchemy error did not occur.

Upon further extensive inspection, it looks like this is simply because `_get_current_trial()` is never called when the worker process isn't running. Do you know if and why this is the case?

**Comment by pmcharrison** [comment]
@jessesnyder that's interesting but confusing..... as far as I know, `_get_current_trial()` should only ever be called from the web process!!

**Comment by jessesnyder** [comment]
I'm vaguely on the scent, but there's still a pretty basic difference between the execution of `dallinger debug` and running flask directly that I can't figure out. 

When I hit the "Do you want to enable custom stimulus and stimulus version filters?" in the context of `dallinger debug`, `_get_current_trail()` is not called. However, when I'm running flask directly (our "develop mode"), it is:
```
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/experiment.py(946)route_response()
-> res = exp.process_response(
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/experiment.py(630)process_response()
-> self.timeline.advance_page(self, participant)
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/timeline.py(1329)advance_page()
-> new_elt.consume(experiment, participant)
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/timeline.py(230)consume()
-> target_elt = self.get_target(experiment, participant)
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/timeline.py(264)get_target()
-> val = call_function(
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/utils.py(107)call_function()
-> return function(*arg_values)
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/trial/main.py(1535)<lambda>()
-> lambda experiment, participant: self._get_current_trial(participant)
  /Users/jesses/projects/py_dev/dlgr_demos/psynet/psynet/trial/main.py(1491)_get_current_trial()
-> return self.trial_class.query.get(trial_id)
```

There's no shortage of indirection in PsyNet here, and I'm having trouble finding where the roads diverge. @pmcharrison If you have any insights about this I'd love to hear them. In `dallinger debug` it seems `_get_current_trial()` is only called after the participant is rejected (for not liking ponies at all, for example).

**Comment by pmcharrison** [comment]
Hi @jessesnyder, finding this quite confusing, as when I run `dallinger debug` I'm finding that `get_current_trial()` is definitely called as soon as I respond to the first question, see video:

https://user-images.githubusercontent.com/22684998/142463156-b1fc6e95-347d-4fe9-99c3-fff405be967a.mp4

**Comment by jessesnyder** [comment]
Looks like I've been the victim of output buffering... I was printing values at that point, and never saw the values. I thought I'd confirmed the problem earlier by raising an exception, but looks like I hadn't (I get the same behavior as your video when I do!) This helps! 🤦🏼‍♂️

**Comment by jessesnyder** [comment]
I feel like I really close, but am missing some bit of insight. @pmcharrison, I'm confident you were correct in your initial suspicion about the problem being rooted in the different fully-qualified class names:
```python
Trial.query.get(trial_id).__class__
# <class 'dallinger_experiment.experiment.AnimalTrial'>

# vs

self.trial_class
# <class 'develop.experiment.AnimalTrial'>
```

When SQLAlchemy wants to instantiate a python object from a database row, it does a check of identity between two classes (and also if one class is a subclass of another) and we fail both checks.

From `_get_current_trial()` into SQLAlchemy object loading:
```
# classes superficially look the same:
sqlalchemy.orm.loading:sub_mapper: mapped class AnimalTrial->info, mapper: mapped class AnimalTrial->info

# but they're not, when their package/namespace is included in the comparison:
sqlalchemy.orm.loading:sub_mapper.class_: <class 'dallinger_experiment.experiment.AnimalTrial'>, mapper.class_: <class 'develop.experiment.AnimalTrial'>

# so False is returned, leading to the exception shortly afterwards
sqlalchemy.orm.loading:return False
```

The problems originate with creating `AnimalTrialMaker` instances in the experiment's `experiment.py` module. In both `dallinger debug` and when running Flask directly ("develop mode"), more than one instance is created, since this happens whenever the module is imported. Also in both cases, sometimes the fully qualified class of `AnimalTrial` that's passed to the `AnimalTrialMaker` constructor has a non-standard (meaning "anything other than `dallinger_experiment.experiment.AnimalTrial`") class name. The difference is in what happens later. 

For some reason, when running `dallinger debug`, the `AnimalTrialMaker` instance on which `_get_current_trial()` is called is always one with the correct `AnimalTrial` class assigned to `self.trial_class`, while when running "develop mode", it's aways an instance with the wrong `AnimalTrial` class.

### dallinger debug
```
# wrong (dallinger cli process)
2021-11-18 12:24:57,223 - root - INFO - In TrialMaker.__init__, trial_class arg: <class 'TEMP_VERIFICATION_PACKAGE.experiment.AnimalTrial'>
2021-11-18 12:24:57,223 - root - INFO - self: <TEMP_VERIFICATION_PACKAGE.experiment.AnimalTrialMaker object at 0x1078aa490>

# right (dallinger cli process)
2021-11-18 12:24:57,229 - root - INFO - In TrialMaker.__init__, trial_class arg: <class 'dallinger_experiment.experiment.AnimalTrial'>
2021-11-18 12:24:57,229 - root - INFO - self: <dallinger_experiment.experiment.AnimalTrialMaker object at 0x11da825e0>

# wrong (web process)
12:25:00 PM web.1   |  INFO:root:In TrialMaker.__init__, trial_class arg: <class 'de8c0017-f5ff-c4ee-925c-0cf5c4b5ecbd.experiment.AnimalTrial'>
12:25:00 PM web.1   |  INFO:root:self: <de8c0017-f5ff-c4ee-925c-0cf5c4b5ecbd.experiment.AnimalTrialMaker object at 0x109bcf070>

# right (web process)
12:25:00 PM web.1   |  INFO:root:In TrialMaker.__init__, trial_class arg: <class 'dallinger_experiment.experiment.AnimalTrial'>
12:25:00 PM web.1   |  INFO:root:self: <dallinger_experiment.experiment.AnimalTrialMaker object at 0x11ff4bf40>

# wrong (worker process)
12:25:02 PM worker.1 |  2021-11-18 12:25:02,258 In TrialMaker.__init__, trial_class arg: <class 'de8c0017-f5ff-c4ee-925c-0cf5c4b5ecbd.experiment.AnimalTrial'>
12:25:02 PM worker.1 |  2021-11-18 12:25:02,259 self: <de8c0017-f5ff-c4ee-925c-0cf5c4b5ecbd.experiment.AnimalTrialMaker object at 0x1200cc610>

# then, the critical moment, when participant clicks that magic button, an AnimalTrialMaker instance with the right class is used
12:25:58 PM web.1   |  INFO:root:IN _get_current_trial, self.trial_class: <class 'dallinger_experiment.experiment.AnimalTrial'>
12:25:58 PM web.1   |  INFO:root:self: <dallinger_experiment.experiment.AnimalTrialMaker object at 0x11ff4bf40>
```

### ./run.sh (with not clock, for simplicity)
```
# wrong
INFO:root:In TrialMaker.__init__, trial_class arg: <class 'develop.experiment.AnimalTrial'>
INFO:root:self: <develop.experiment.AnimalTrialMaker object at 0x10a4c3820>

# wrong
INFO:root:In TrialMaker.__init__, trial_class arg: <class 'develop.experiment.AnimalTrial'>
INFO:root:self: <develop.experiment.AnimalTrialMaker object at 0x10d18a520>

# right
INFO:root:In TrialMaker.__init__, trial_class arg: <class 'dallinger_experiment.experiment.AnimalTrial'>
INFO:root:self: <dallinger_experiment.experiment.AnimalTrialMaker object at 0x1234ce310>

# but in _get_current_trial, the wrong AnimalTrialMaker is used
INFO:root:IN _get_current_trial, self.trial_class: <class 'develop.experiment.AnimalTrial'>
INFO:root:self: <develop.experiment.AnimalTrialMaker object at 0x10d18a520>
```

Taking a break and hoping lightning will strike.

**Comment by pmcharrison** [comment]
Now fixed in #3929.
